test/e2e/specs/site-editor/block-style-variations.spec.js

```diff
@@ -0,0 +1,303 @@
+/**
+ * WordPress dependencies
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+const TEST_PAGE_TITLE = 'Test Page for Block Style Variations';
+test.use( {
+	siteEditorBlockStyleVariations: async ( { page, editor }, use ) => {
+		await use( new SiteEditorBlockStyleVariations( { page, editor } ) );
+	},
+} );
+
+test.describe( 'Block Style Variations', () => {
+	let stylesPostId;
+	test.beforeAll( async ( { requestUtils } ) => {
+		await Promise.all( [
+			requestUtils.activateTheme(
+				'gutenberg-test-themes/style-variations'
+			),
+			requestUtils.deleteAllPages(),
+		] );
+		stylesPostId = await requestUtils.getCurrentThemeGlobalStylesPostId();
+	} );
+
+	test.afterAll( async ( { requestUtils } ) => {
+		await Promise.all( [
+			requestUtils.activateTheme( 'twentytwentyone' ),
+			requestUtils.deleteAllPages(),
+		] );
+	} );
+
+	test.beforeEach( async ( { requestUtils, admin } ) => {
+		await Promise.all( [
+			requestUtils.deleteAllPages(),
+			admin.visitSiteEditor(),
+		] );
+	} );
+
+	test( 'apply block styles variations to nested blocks', async ( {
+		editor,
+		page,
+	} ) => {
+		await draftNewPage( page );
+		await addPageContent( editor, page );
+		const firstGroup = editor.canvas
+			.locator( '[data-type="core/group"]' )
+			.first();
+		const secondGroup = editor.canvas
+			.locator( '[data-type="core/group"]' )
+			.nth( 1 );
+		const thirdGroup = editor.canvas
+			.locator( '[data-type="core/group"]' )
+			.nth( 2 );
+
+		// Apply a block style to the parent Group block.
+		await editor.selectBlocks( firstGroup );
+		await editor.openDocumentSettingsSidebar();
+		await page.getByRole( 'tab', { name: 'Styles' } ).click();
+		await page
+			.getByRole( 'button', { name: 'Block Style Variation A' } )
+			.click( { force: true } );
+
+		// Check parent styles.
+		await expect( firstGroup ).toHaveCSS( 'border-style', 'dotted' );
+		await expect( firstGroup ).toHaveCSS( 'border-width', '1px' );
+
+		// Check nested child and grandchild group block inherited styles.
+		await expect( secondGroup ).toHaveCSS( 'border-style', 'solid' );
+		await expect( secondGroup ).toHaveCSS( 'border-width', '11px' );
+		await expect( thirdGroup ).toHaveCSS( 'border-style', 'solid' );
+		await expect( thirdGroup ).toHaveCSS( 'border-width', '11px' );
+
+		// Apply a block style to the first, nested Group block.
+		await editor.selectBlocks( secondGroup );
+		await page.getByRole( 'tab', { name: 'Styles' } ).click();
+		await page
+			.getByRole( 'button', { name: 'Block Style Variation B' } )
+			.click( { force: true } );
+
+		// Check nested child styles.
+		await expect( secondGroup ).toHaveCSS( 'border-style', 'dashed' );
+		await expect( secondGroup ).toHaveCSS( 'border-width', '2px' );
+
+		// Check nested grandchild group block inherited styles.
+		await expect( thirdGroup ).toHaveCSS( 'border-style', 'groove' );
+		await expect( thirdGroup ).toHaveCSS( 'border-width', '22px' );
```

Comment:
I know this test is a beast already but I was wondering if we should also apply the first variation to the grandchild here at the end?

Something to ensure that earlier defined style variations don't have their styles overridden by a later definition's inner block type styles.

Reply:
Just checking I understand - so the test is to apply a block style variation (A) to the nested grandchild block to ensure that A's styles overwrite any inherited styles from parent blocks.

I'll add it and you can tell me if I got it right 😄

Reply:
Yep, I believe we're on the same page.

If only a simple stylesheet was being generated and there was a single set of CSS for each variation, including inner block type styles, for example:

```css
:root :where(.is-style-variation-a .wp-block-group) {...}
:root :where(.is-style-variation-b .wp-block-group) {...}
```

Then checking that variation B gets applied when selecting it on the child covers whether that variation then overcomes the inner block type styles for variation A applied to the parent. So far, so good 👍 

Given the same specificity in the example above, the variation B inner block type styles would override the selection of variation A on the grandchild due to being loaded later. 

Adding this extra test for the application of variation A again on the grandchild will cover that we're not relying on the load order of the CSS. In other words, it ensures we're creating the per-instance styles appropriately so that the most immediate variation's styles are in effect.

I hope that makes it clearer 🤞